src/interfaces/highs_c_api.cpp

```diff
@@ -84,18 +92,124 @@ HighsInt Highs_mipCall(const HighsInt numcol, const HighsInt numrow,
+    const double* rowupper, const HighsInt* astart, const HighsInt* aindex,
+    const double* avalue, const HighsInt* qstart, const HighsInt* qindex,
+    const double* qvalue, double* colvalue, double* coldual, double* rowvalue,
+    double* rowdual, HighsInt* colbasisstatus, HighsInt* rowbasisstatus,
+    HighsInt* modelstatus) {
+  Highs highs;
+  highs.setOptionValue("output_flag", false);
+  HighsStatus status =
+      highs.passModel(numcol, numrow, numnz, q_numnz, a_format, q_format, sense,
+                      offset, colcost, collower, colupper, rowlower, rowupper,
+                      astart, aindex, avalue, qstart, qindex, qvalue);
+  if (status != HighsStatus::kOk) return (HighsInt)status;
+
+  status = highs.run();
+
+  if (status == HighsStatus::kOk) {
+    HighsSolution solution;
+    HighsBasis basis;
+    solution = highs.getSolution();
+    basis = highs.getBasis();
+    *modelstatus = (HighsInt)highs.getModelStatus();
+    const HighsInfo& info = highs.getInfo();
+
+    const bool copy_col_value =
+        colvalue != NULL &&
+        info.primal_solution_status != SolutionStatus::kSolutionStatusNone;
+    const bool copy_col_dual =
+        coldual != NULL &&
+        info.dual_solution_status != SolutionStatus::kSolutionStatusNone;
+    const bool copy_col_basis = colbasisstatus != NULL && basis.valid;
+    for (HighsInt i = 0; i < numcol; i++) {
+      if (copy_col_value) colvalue[i] = solution.col_value[i];
+      if (copy_col_dual) coldual[i] = solution.col_dual[i];
+      if (copy_col_basis) colbasisstatus[i] = (HighsInt)basis.col_status[i];
+    }
+
+    const bool copy_row_value =
+        rowvalue != NULL &&
+        info.primal_solution_status != SolutionStatus::kSolutionStatusNone;
+    const bool copy_row_dual =
+        rowdual != NULL &&
+        info.dual_solution_status != SolutionStatus::kSolutionStatusNone;
+    const bool copy_row_basis = rowbasisstatus != NULL && basis.valid;
+    for (HighsInt i = 0; i < numrow; i++) {
+      if (copy_row_value) rowvalue[i] = solution.row_value[i];
+      if (copy_row_dual) rowdual[i] = solution.row_dual[i];
+      if (copy_row_basis) rowbasisstatus[i] = (HighsInt)basis.row_status[i];
+    }
+  }
+  return (HighsInt)status;
+}
+
+HighsInt Highs_lpDimMpsRead(
+    //			    const char* filename,
+    HighsInt* numcol, HighsInt* numrow, HighsInt* numnz) {
+  Highs highs;
+  highs.setOptionValue("output_flag", false);
+  const char* filename = "ml.mps";
+  HighsStatus status = highs.readModel(filename);
+  const HighsLp& lp = highs.getLp();
+  *numcol = lp.numCol_;
+  *numrow = lp.numRow_;
+  *numnz = lp.Astart_[lp.numCol_];
+  return (HighsInt)status;
+}
+
+HighsInt Highs_lpDataMpsRead(
+    //			     const char* filename,
+    const HighsInt numcol, const HighsInt numrow, HighsInt* sense,
+    double* offset, double* colcost, double* collower, double* colupper,
+    double* rowlower, double* rowupper, HighsInt* astart, HighsInt* aindex,
+    double* avalue) {
```

Comment:
Rather than having these two separate methods, `lpDataMpsRead` and `lpDimMpsRead`, the Gurobi approach is to only have the `Data` version, and allow `NULL` to be passed as the vectors. 

Then the approach is:
 - call it will `NULL` and obtain `numcol`, `numrow`, and `numnz`
 - allocate vectors based on the size
 - call it again with the allocated vectors

Thoughts on that approach?

Reply:
I'm sure that there's a better approach. I had to create something to allow an MSc student to read MPS files into Python! 
Passing the file name as a string from Python was beyond me, but it's clearly not so hard in C!

Reply:
> Passing the file name as a string from Python was beyond me

Yeah I'm no help on that front, unfortunately.

Although a quick google suggests: https://nesi.github.io/perf-training/python-scatter/ctypes
<img width="476" alt="image" src="https://user-images.githubusercontent.com/8177701/125532817-fd82b9e1-fd37-4372-83a8-eb45c3f7dc0c.png">

Reply:
Thanks. We lack local knowledge of Python, and it's an important interface, even if we're in SciPy!

Reply:
I don't know everything about ctypes, but happy to help out if I can.  I've seen more and more requests for modeling package wrappers for HiGHS, but it's just not something we're interested in for SciPy (we're not that kind of package).

For string encoding, you can also use a more convenient syntax:
```python
b'myCoolFile.mps' == 'myCoolFile.mps'.encode()
```

@jajhall You might also be interested in some comparisons we've been doing for LP solvers in Python-land for a SciPy conference.  I present tomorrow, I will share slides with you afterward.  We're comparing against Python interfaces for Google's GLOP, SCIP, CLP, and others

Reply:
> I present tomorrow, I will share slides with you afterward. We're comparing against Python interfaces for Google's GLOP, SCIP, CLP, and others

I'd also be very interested in this.